packages/cosmic-swingset/lib/ag-solo/vats/offer-compiler.js

```diff
@@ -0,0 +1,213 @@
+      assert(
+        purse,
+        details`Want role ${role} pursePetname ${amount.pursePetname} is not a purse`,
+      );
+      rolePurses[role] = purse;
+      setPurseAmount(roleOfferRules.want, role, purse, amount.extent);
+    });
+
+    if (offerRulesTemplate.offer) {
+      roleOfferRules.offer = {};
+    }
+    Object.entries(offerRulesTemplate.offer || {}).forEach(([role, amount]) => {
+      assert(
+        amount.pursePetname,
+        details`Offer role ${role} has no pursePetname`,
+      );
+      const purse = petnameToPurse.get(amount.pursePetname);
+      assert(
+        purse,
+        details`Offer role ${role} pursePetname ${amount.pursePetname} is not a purse`,
+      );
+      rolePurses[role] = purse;
+      setPurseAmount(roleOfferRules.offer, role, purse, amount.extent);
+    });
+
+    return { roleOfferRules, rolePurses };
+  }
+
+  const { roleOfferRules, rolePurses } = createRoleOfferRulesAndPurses(
+    offerRulesTemplate,
+  );
+
+  // Enrich the offerRulesTemplate.
+  const newOfferRulesTemplate = { ...offerRulesTemplate };
+  if (offerRulesTemplate.want) {
+    const newRules = {};
+    Object.entries(offerRulesTemplate.want || {}).forEach(([role, amount]) => {
+      newRules[role] = { ...amount, ...roleIssuerNames[role] };
+    });
+    newOfferRulesTemplate.want = newRules;
+  }
+
+  if (offerRulesTemplate.offer) {
+    const newRules = {};
+    Object.entries(offerRulesTemplate.offer || {}).forEach(([role, amount]) => {
+      newRules[role] = { ...amount, ...roleIssuerNames[role] };
+    });
+    newOfferRulesTemplate.offer = newRules;
+  }
+
+  // Resave the enriched offerDesc.
+  idToOfferDesc.set(id, {
+    ...offerDesc,
+    offerRulesTemplate: newOfferRulesTemplate,
+  });
+
+  // Get the instance.
+  const instanceHandle = await E(registrar).get(instanceRegKey);
+  const {
+    publicAPI,
+    roles: contractRoleIssuers,
+    terms: { issuers: contractIssuers },
+  } = await E(zoe).getInstance(instanceHandle);
+
+  const roleIssuers = { ...contractRoleIssuers };
+  Object.values(contractIssuerIndexToRole).forEach((role, i) => {
+    roleIssuers[role] = contractIssuers[i];
+  });
+
+  async function finishCompile(offerRules, purses) {
+    const roleBrands = {};
+    let cachedRoleBrandsP;
+    const getRoleBrandsP = () => {
+      if (cachedRoleBrandsP) {
+        return cachedRoleBrandsP;
+      }
+      cachedRoleBrandsP = Promise.all(
+        Object.entries(roleIssuers).map(async ([role, issuer]) => {
+          roleBrands[role] = await E(issuer).getBrand();
+        }),
+      );
+      return cachedRoleBrandsP;
+    };
+
+    const replacePlaceholderRoles = async (roles, role) => {
+      if (role[0] !== '$') {
+        return;
+      }
+
+      // It's a placeholder role.  Find the first matching role with that brand.
+      await getRoleBrandsP();
+      const { brand } = roles[role];
+      const roleEnt = Object.entries(roleBrands).find(
```

Comment:
I've added assertions to prevent aliasing when expanding "multiroles" (ones that end in a star).  Ordinary roles will never alias unless explicitly requested that way.